contracts/MaticX.sol

```diff
@@ -1,386 +1,392 @@
 contract MaticX is
 	IMaticX,
 	ERC20Upgradeable,
 	AccessControlUpgradeable,
 	PausableUpgradeable
 {
 	using SafeERC20Upgradeable for IERC20Upgradeable;
+	using StringsUpgradeable for string;
+
+	bytes32 public constant PREDICATE_ROLE = keccak256("PREDICATE_ROLE");
+	bytes32 public constant BOT = keccak256("BOT");
+	uint256 private constant NOT_ENTERED = 1;
+	uint256 private constant ENTERED = 2;
 
 	address private validatorRegistry;
 	address private stakeManager;
-	address private polygonERC20;
+	address private maticToken;
 
 	address public override treasury;
 	string public override version;
 	uint8 public override feePercent;
 
-	bytes32 public constant INSTANT_POOL_OWNER = keccak256("IPO");
+	address private instantPoolOwner_deprecated;
+	uint256 private instantPoolMatic_deprecated;
+	uint256 private instantPoolMaticX_deprecated;
 
-	address public override instantPoolOwner;
-	uint256 public override instantPoolMatic;
-	uint256 public override instantPoolMaticX;
-
-	/// @notice Mapping of all user ids with withdraw requests.
 	mapping(address => WithdrawalRequest[]) private userWithdrawalRequests;
+	address public override fxStateRootTunnel;
+	address private polToken;
+	uint256 private reentrancyGuardStatus;
 
-	bytes32 public constant PREDICATE_ROLE = keccak256("PREDICATE_ROLE");
+	/// ------------------------------ Modifiers -------------------------------
 
-	address public override fxStateRootTunnel;
+	/// @notice Enables guard from reentrant calls.
+	modifier nonReentrant() {
+		require(
+			reentrancyGuardStatus != ENTERED,
+			"ReentrancyGuard: reentrant call"
+		);
+		reentrancyGuardStatus = ENTERED;
+		_;
+		reentrancyGuardStatus = NOT_ENTERED;
+	}
 
-	bytes32 public constant BOT = keccak256("BOT");
+	/// -------------------------- Initializers --------------------------------
+
+	/// @dev The constructor is disabled for a proxy upgrade.
+	/// @custom:oz-upgrades-unsafe-allow constructor
+	constructor() {
+		_disableInitializers();
+	}
 
-	/// @notice Initialize the MaticX contract.
+	/// @notice Initializes the current contract.
 	/// @param _validatorRegistry - Address of the validator registry
 	/// @param _stakeManager - Address of the stake manager
-	/// @param _polygonERC20 - Address of matic token on Ethereum
+	/// @param _maticToken - Address of the Matic token
 	/// @param _manager - Address of the manager
-	/// @param _instantPoolOwner - Address of the instant pool owner
 	/// @param _treasury - Address of the treasury
 	function initialize(
 		address _validatorRegistry,
 		address _stakeManager,
-		address _polygonERC20,
+		address _maticToken,
 		address _manager,
-		address _instantPoolOwner,
 		address _treasury
-	) external override initializer {
-		__AccessControl_init();
-		__Pausable_init();
-		__ERC20_init("Liquid Staking Matic", "MaticX");
-
-		_setupRole(DEFAULT_ADMIN_ROLE, _manager);
-		_setupRole(INSTANT_POOL_OWNER, _instantPoolOwner);
-		instantPoolOwner = _instantPoolOwner;
+	) external initializer {
```

Comment:
```suggestion
	) external reinitializer(2) {
```

Reply:
No, we don't plan to call the initialize again. OZ contracts offers a mechanism of contract reinitialization like that https://docs.openzeppelin.com/contracts/5.x/api/proxy#Initializable-reinitializer-uint64-. So the idea is just to reinitialize the contract with new new required data.